### PR TITLE
[SYCLomatic] Enable migration of binary, reduction, batchnorm, norm, convolution layer

### PR DESCRIPTION
**Pull request by intwanghao**

### CODE REVIEW

File: clang/lib/DPCT/DNNAPIMigration.cpp [hunk 1]
```diff
@@ -76,17 +82,23 @@ void CuDNNTypeRule::runRule(const MatchFinder::MatchResult &Result) {
...
-             "data type double");
-      return;
+
+    if (EnumName.find("CUDNN_STATUS_") != std::string::npos) {
+      if (auto EC = dyn_cast<EnumConstantDecl>(E->getDecl())) {
+        std::string Repl = toString(EC->getInitVal(), 10);
+        emplaceTransformation(new ReplaceStmt(E, Repl));
+        return;
+      }
+    } else if(EnumName == "CUDNN_BATCHNORM_SPATIAL_PERSISTENT") {
```
Comment (ziranzha) [inline clang/lib/DPCT/DNNAPIMigration.cpp:95]: Should add lit test for this warning

File: clang/test/dpct/helper_files_ref/include/dnnl_utils.hpp [hunk 2]
```diff
@@ -408,158 +724,680 @@ class engine_ext {
...
-  void activation_backward(activation_desc &desc, float alpha,
+  /// Retrieving the user's SYCL queue set in the oneDNN engine.
+  /// \returns Pointer to the SYCL queue.
+  sycl::queue *get_queue() const { return _q; }
+  /// Setting all elements of a memory to a given value.
+  /// \param [in] src_desc Source memory descriptor.
+  /// \param [in] src Pointer to source data.
+  /// \param [in] valuePtr Pointer to a signle value.
+  /// \returns An event representing the fill operations.
+  sycl::event fill(const memory_desc_ext &src_desc, void *src,
```
Comment (zhimingwang36) [inline clang/test/dpct/helper_files_ref/include/dnnl_utils.hpp]: here break backward compatibility,  sugguest to keep the void fill() interface.

File: clang/test/dpct/helper_files_ref/include/dnnl_utils.hpp [hunk 3]
```diff
@@ -408,158 +724,680 @@ class engine_ext {
...
+  /// \param [in] alpha Value to scaling factors used to scale the computed
+  /// value.
+  /// \param [in] src_desc Source memory descriptor.
+  /// \param [in] src Pointer to source data.
+  /// \param [in] beta Value to scaling factors used to scale the prior value
+  /// in the destination memory.
+  /// \param [in] dst_desc Destination memory descriptor.
+  /// \param [out] dst Pointer to destination data.
+  /// \returns An event representing the reorder operations.
+  sycl::event reorder(float alpha, const memory_desc_ext &src_desc, void *src,
```
Comment (zhimingwang36) [inline clang/test/dpct/helper_files_ref/include/dnnl_utils.hpp]: the change break the backward compatibility,  let's keep both async and sync interface.

File: clang/test/dpct/helper_files_ref/include/dnnl_utils.hpp [hunk 4]
```diff
@@ -695,16 +1555,104 @@ void memory_desc_ext::get(int requested_ndims, dpct::library_data_t *dt, int *nd
...
+    transform_no_zero_with_type<int32_t>(_q, src, dst, element_num);
+    break;
+  case ::dnnl::memory::data_type::s8:
+    transform_no_zero_with_type<int8_t>(_q, src, dst, element_num);
+    break;
+  case ::dnnl::memory::data_type::u8:
+    transform_no_zero_with_type<uint8_t>(_q, src, dst, element_num);
+    break;
+  default:
+    throw std::runtime_error("fill: unsupported data type.");
```
Comment (zhimingwang36) [inline clang/test/dpct/helper_files_ref/include/dnnl_utils.hpp]: message in L1604:  fill ==> transform_no_zero

File: clang/test/dpct/helper_files_ref/include/dnnl_utils.hpp [hunk 5]
```diff
@@ -29,77 +33,126 @@ class memory_desc_ext {
...
+  /// stride of each memory dimension.
+  void get(int requested_ndims, dpct::library_data_t *dt, int *ndims,
+           int dims[], int strides[]) const;
+  /// Getting parameters from a ND memory.
+  /// \param [in] requested_ndims Requested number of dimensions to get from a
+  /// given memory descriptor.
+  /// \param [out] dt Data type.
+  /// \param [out] tag Format tag.
+  /// \param [out] ndims Dimension of the memory.
+  /// \param [out] dims  Array of dimension requested_ndims that contain the
```
Comment (zhimingwang36) [inline clang/test/dpct/helper_files_ref/include/dnnl_utils.hpp]: remove the extra emtpy in L129 L401, L555,

File: clang/test/dpct/helper_files_ref/include/dnnl_utils.hpp [hunk 6]
```diff
@@ -408,158 +724,680 @@ class engine_ext {
...
-/// \param [in] diff_src_desc Differential source memory descriptor.
-/// \param [out] diff_src Pointer to differential source data.
-  void activation_backward(activation_desc &desc, float alpha,
+  /// Retrieving the user's SYCL queue set in the oneDNN engine.
+  /// \returns Pointer to the SYCL queue.
+  sycl::queue *get_queue() const { return _q; }
+  /// Setting all elements of a memory to a given value.
+  /// \param [in] src_desc Source memory descriptor.
+  /// \param [in] src Pointer to source data.
+  /// \param [in] valuePtr Pointer to a signle value.
```
Comment (zhimingwang36) [inline clang/test/dpct/helper_files_ref/include/dnnl_utils.hpp]: signle ==> single.

File: clang/test/dpct/helper_files_ref/include/dnnl_utils.hpp [hunk 7]
```diff
@@ -408,158 +724,680 @@ class engine_ext {
...
-                    void *dst, const memory_desc_ext &diff_dst_desc,
-                    void *diff_dst, const memory_desc_ext &src_desc, void *src,
-                    float beta, const memory_desc_ext &diff_src_desc,
-                    void *diff_src, ::dnnl::memory *workspace = nullptr);
+                           const memory_desc_ext &src_desc, void *src,
+                           float beta, const memory_desc_ext &diff_src_desc,
+                           void *diff_src, ::dnnl::memory *workspace = nullptr);
+
+  /// Derives a memory descriptor for the batch normalization scale, bias, mean,
+  /// variance from the source memory descripotr and batch normalization mode.
```
Comment (zhimingwang36) [inline clang/test/dpct/helper_files_ref/include/dnnl_utils.hpp]: descripotr==> descriptor

File: clang/test/dpct/helper_files_ref/include/dnnl_utils.hpp [hunk 8]
```diff
@@ -408,158 +724,680 @@ class engine_ext {
...
+  /// \param [out] desc Derived memory descriptor.
+  /// \param [in] src_desc Source memory descriptor.
+  /// \param [in] mode Batch normalization mode.
+  static void derive_batch_normalization_memory_desc(memory_desc_ext &desc,
+                                              const memory_desc_ext &src_desc,
+                                              batch_normalization_mode mode);
+
+  /// Derives a memory descriptor for the batch normalization scale, bias, mean,
+  /// variance from the source memory descripotr and batch normalization mode.
```
Comment (zhimingwang36) [inline clang/test/dpct/helper_files_ref/include/dnnl_utils.hpp]: descripotr ==> descriptor.

File: clang/test/dpct/helper_files_ref/include/dnnl_utils.hpp [hunk 9]
```diff
@@ -408,158 +724,680 @@ class engine_ext {
...
+  static void derive_batch_normalization_memory_desc(memory_desc_ext &scale_bias_desc,
+                                             memory_desc_ext &mean_var_desc,
+                                             const memory_desc_ext &src_desc,
+                                             batch_normalization_mode mode);
+
+  /// Get the size of workspace that needed by batch normalization. The data stored
+  /// in workspace must be preserved between forward and backward.
+  /// \param [in] ops Batch normalization operation mode. This mode can set to
+  /// perform only batch normalization, or batch normalization followed by
+  /// activation, or batch normalization followed by element-wise addtion and
```
Comment (zhimingwang36) [inline clang/test/dpct/helper_files_ref/include/dnnl_utils.hpp]: addtion  ==> addition

File: clang/test/dpct/helper_files_ref/include/dnnl_utils.hpp [hunk 10]
```diff
@@ -408,158 +724,680 @@ class engine_ext {
...
+  /// activation.
+  /// \param [in] src_desc Source memory descriptor.
+  /// \returns Size of workpsace.
```
Comment (zhimingwang36) [inline clang/test/dpct/helper_files_ref/include/dnnl_utils.hpp]: workpsace ==> workspace.

File: clang/test/dpct/helper_files_ref/include/dnnl_utils.hpp [hunk 11]
```diff
@@ -408,158 +724,680 @@ class engine_ext {
...
+      const memory_desc_ext &src_desc, void *src, float beta,
+      const memory_desc_ext &dst_desc, void *dst,
+      const memory_desc_ext &scale_bias_mean_var_desc, void *scale, void *bias,
+      void *mean, void *var);
+
+  /// Computing a specified batch normalization inference stage function value.
+  /// \param [in] mode Batch normalization mode.
+  /// \param [in] ops Batch normalization operation mode. This mode can set to
+  /// perform only batch normalization, or batch normalization followed by
+  /// activation, or batch normalization followed by element-wise addtion and
```
Comment (zhimingwang36) [inline clang/test/dpct/helper_files_ref/include/dnnl_utils.hpp]: addtion ==> addition.  Also,  L1081,  L1123, L1199, L1249,

File: clang/test/dpct/helper_files_ref/include/dnnl_utils.hpp [hunk 12]
```diff
@@ -714,11 +1662,71 @@ engine_ext::compress_spatial_dimensions_to_channel(::dnnl::memory::desc desc) {
...
+  auto dims = desc.dims();
+  if (ndims == 4) {
+    return ::dnnl::memory::desc(dims, desc.data_type(),
+                                ::dnnl::memory::format_tag::nchw);
+  }
+  return ::dnnl::memory::desc(dims, desc.data_type(),
+                              ::dnnl::memory::format_tag::ncdhw);
+}
+
+/// If the alpha = 0 and beta = 1, then the destination(dst = alpha * out +
```
Comment (zhimingwang36) [inline clang/test/dpct/helper_files_ref/include/dnnl_utils.hpp]: insert a empty before "(dst"

File: clang/test/dpct/helper_files_ref/include/dnnl_utils.hpp [hunk 13]
```diff
@@ -714,11 +1662,71 @@ engine_ext::compress_spatial_dimensions_to_channel(::dnnl::memory::desc desc) {
...
+/// beta * prior_dst) have no change. In this case this funciton return true
```
Comment (zhimingwang36) [inline clang/test/dpct/helper_files_ref/include/dnnl_utils.hpp]: funciton ==> function.

File: clang/test/dpct/helper_files_ref/include/dnnl_utils.hpp [hunk 14]
```diff
@@ -29,77 +33,126 @@ class memory_desc_ext {
...
+  /// \param [out] dt Data type.
+  /// \param [out] tag Format tag.
+  /// \param [out] ndims Dimension of the memory.
+  /// \param [out] dims Array of dimension requested_ndims that contain the 
+  /// size of each memory dimension.
+  void get(int requested_ndims, dpct::library_data_t *dt,
+           memory_format_tag *tag, int *ndims, int dims[]) const;
+  /// Getting dims from a ND memory.
+  /// \return The dims.
+  std::vector<int64_t> get_dims() const { return _desc.dims(); }
```
Comment (zhimingwang36) [inline clang/test/dpct/helper_files_ref/include/dnnl_utils.hpp:135]: return type change to :  const std::vector<> & ,  avoid copy.

File: clang/test/dpct/helper_files_ref/include/dnnl_utils.hpp [hunk 15]
```diff
@@ -340,62 +413,305 @@ class pooling_desc {
...
+  /// \param [out] ndims Dimension of the pooling operation.
+  /// \param [out] paddings Array of dimension ndims containing the padding size
+  /// of each dimension.
+  /// \param [out] strides Array of dimension ndims containing the stride size of
+  /// each dimension.
+  /// \param [out] dilates Array of dimension ndims containing the dilate size of
+  /// each dimension.
+  void get(int requested_ndims, int *ndims, int paddings[], int strides[],
+           int dilates[]) const {
+    *ndims = _strides.size();
```
Comment (zhimingwang36) [inline clang/test/dpct/helper_files_ref/include/dnnl_utils.hpp:509]: add a asset(requested_ndims <=ndims).

File: clang/test/dpct/helper_files_ref/include/dnnl_utils.hpp [hunk 16]
```diff
@@ -259,74 +327,79 @@ class pooling_desc {
...
+  /// \param [out] kernel Array of dimension ndims containing the kernel size of
+  /// each dimension.
+  /// \param [out] padding Array of dimension ndims containing the padding size
+  /// of each dimension.
+  /// \param [out] stride Array of dimension ndims containing the stride size of
+  /// each dimension.
+  void get(int requested_ndims, ::dnnl::algorithm *alg, int *ndims,
+           int kernel[], int padding[], int stride[]) const {
     *alg = _alg;
+    *ndims = _stride.size();
```
Comment (zhimingwang36) [inline clang/test/dpct/helper_files_ref/include/dnnl_utils.hpp:344]: add  assert ()

File: clang/test/dpct/helper_files_ref/include/dnnl_utils.hpp [hunk 17]
```diff
@@ -408,158 +724,869 @@ class engine_ext {
...
+  /// \param [in] summand_desc Summand memory descriptor.
+  /// \param [in] summand Pointer to summand data.
+  /// \param [in] scale_bias_desc Scale, bias memory descriptor.
+  /// \param [in] scale Pointer to scale data.
+  /// \param [in] bias Pointer to bias data.
+  /// \param [in] mean_var_desc Mean, variance memory descriptor.
+  /// \param [in] mean Pointer to mean data.
+  /// \param [in] var Pointer to variance data.
+  /// \returns An event representing the batch normalization forward operations.
+  sycl::event async_batch_normalization_forward_inference_ex(
```
Comment (zhimingwang36) [inline clang/test/dpct/helper_files_ref/include/dnnl_utils.hpp]: how about we use async_batch_normalization_forward_inference() ?

File: clang/test/dpct/helper_files_ref/include/dnnl_utils.hpp [hunk 18]
```diff
@@ -408,158 +724,869 @@ class engine_ext {
...
+  /// \param [in] bias Pointer to bias data.
+  /// \param [out] running_mean Pointer to running mean data.
+  /// \param [out] running_var Pointer to running variance data.
+  /// \param [out] saved_mean Pointer to optional cache to save mean data.
+  /// \param [out] saved_var Pointer to optional cache to save variance data.
+  /// \param [in] workspace_size Size of workspace.
+  /// \param [out] workspace Pointer to workspace generated from forward
+  /// propagation.
+  /// \returns An event representing the batch normalization forward operations.
+  sycl::event async_batch_normalization_forward_training_ex(
```
Comment (zhimingwang36) [inline clang/test/dpct/helper_files_ref/include/dnnl_utils.hpp]: let' s remove _ex postfix.

File: clang/test/dpct/helper_files_ref/include/dnnl_utils.hpp [hunk 19]
```diff
@@ -714,11 +1851,71 @@ engine_ext::compress_spatial_dimensions_to_channel(::dnnl::memory::desc desc) {
...
+  if (ndims == 4) {
+    return ::dnnl::memory::desc(dims, desc.data_type(),
+                                ::dnnl::memory::format_tag::nchw);
+  }
+  return ::dnnl::memory::desc(dims, desc.data_type(),
+                              ::dnnl::memory::format_tag::ncdhw);
+}
+
+/// If the alpha = 0 and beta = 1, then the destination (dst = alpha * out +
+/// beta * prior_dst) have no change. In this case this function return true
```
Comment (zhimingwang36) [inline clang/test/dpct/helper_files_ref/include/dnnl_utils.hpp]: return   ==> returns.

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Review by zhimingwang36 — APPROVED** [review]
LGTM  (I'll merged it as previoius test result the PRE-CI passed).

**Pull request merged**